docs/lang/articles/advanced/differentiable_programming.md

```diff
@@ -169,69 +156,214 @@ start up.
 
 :::tip
 See
-[examples/mpm_lagrangian_forces.py](https://github.com/taichi-dev/taichi/blob/master/examples/mpm_lagrangian_forces.py)
+[examples/simulation/mpm_lagrangian_forces.py](https://github.com/taichi-dev/taichi/blob/master/examples/simulation/mpm_lagrangian_forces.py)
 and
-[examples/fem99.py](https://github.com/taichi-dev/taichi/blob/master/examples/fem99.py)
+[examples/simulation/fem99.py](https://github.com/taichi-dev/taichi/blob/master/examples/simulation/fem99.py)
 for examples on using autodiff for MPM and FEM.
 :::
 
 ## Using `kernel.grad()`
 
-TODO: Documentation WIP.
+As mentioned above, `ti.Tape()` can only track a 0D field as output variable.
+If there're multiple output variables that you want to back-propagate
+gradients to inputs, `kernel.grad()` should be used instead of `ti.Tape()`.
 
-## Kernel Simplicity Rule
+```python {13-14}
+import taichi as ti
+ti.init()
+
+N = 16
+
+x = ti.field(dtype=ti.f32, shape=N, needs_grad=True)
+loss = ti.field(dtype=ti.f32, shape=(), needs_grad=True)
+loss2 = ti.field(dtype=ti.f32, shape=(), needs_grad=True)
+
+@ti.kernel
+def func():
+    for i in x:
+       loss[None] += x[i] ** 2
+       loss2[None] += x[i]
+
+for i in range(N):
+    x[i] = i
+loss.grad[None] = 1
+loss2.grad[None] = 1
+
+func()
+func.grad()
+for i in range(N):
+    assert x.grad[i] == i * 2 + 1
+```
+
+:::tip
+It might be tedious to write out `need_grad=True` for every input in a complicated use case.
+Alternatively Taichi provides `ti.root.lazy_grad()` API that can automatically place the
+adjoint fields following the layout of their primal fields.
```

Comment:
```suggestion
gradient fields following the layout of their primal fields.
```
"Gradient" may be a better choice over the mathematically more accurate "adjoint" if we don't expect most readers have backgrounds in automatic differentiation paradigms.